RFC.md

```diff
@@ -0,0 +1,466 @@
+
+# Do inference with the model the same as a text model
+# Fusion models have the same signatures as TransformerDeocder
+>>> data = flamingo_transform(...) # Message formated text + images
+>>> model(**data)
+tensor([-1.0885, ..., .0231], grad_fn=<ViewBackward0>)
+```
+
+> [!Note]
+> This RFC proposes a complete design for supporting fusion models, but not all of the design is required to be built to support any one given fusion style model. This RFC gives us an opportunity to take a holistic view over the space, but it is recommended that each module/component is built as need.
+### Fusion Models
+DeepFusion requires that Decoder is already modified to have any additional layers for accepting encoder inputs (these should be wrapped with FusionLayer to be able to track these additional parameters and work with checkpoints). For inference all of the cache can be managed by the decoder, so these methods are simply passed through.
+
+```python
+class DeepFusionModel(nn.Module):
+	def __init__(self,
+		decoder: TransformerDecoder,
+		encoder: nn.Module,
+		...
+	):
+
+	def setup_caches(self, batch_size: int, dtype: torch.dtype) -> None:
+        """Setup key value caches for attention calculation.
+
+        Args:
+            batch_size (int): batch size for the caches.
+            dtype (torch.dtype): dtype for the caches.
+        """
+        self.decoder.setup_caches(batch_size, dtype)
+
+	def caches_are_enabled(self) -> bool:
+        """Check if the key value caches are setup."""
+        return self.decoder.caches_are_enabled()
+
+	def reset_caches(self):
+        """Reset the key value caches."""
+        self.decoder.reset_caches()
+
+	def forward(
+		self,
+        tokens: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+	if 	encoder_input is not None:
+		# during decoding, the output of the encoder
+		# is cached in the decoder cross attention kv_cache.
+		# so encoder_input is only called when there's a new image
+		encoder_embed = self.encoder(**encoder_input)
+	# input_pos is tracked by kv_cache now, but left in for
+	# now to support sample packing and can be used here
+	if input_pos is not None and encoder_mask is not None:
+		encoder_mask = encoder_mask[input_pos]
+	output = self.decoder(tokens, mask, encoder_embed, encoder_mask, input_pos)
+	return output
+
+ ```
+
+EarlyFusion is more complicated from the fusion side, but does not require a special Decoder. EarlyFusion needs to call `decoder.tok_embeddings` directly so it can merge the embeddings with the encoded embeddings. As to not modify the `state_dict` it employs two hooks to keep `tok_embeddings` inside the decoder for the `state_dict`.
+
+```python
+class EarlyFusionModel(nn.Module):
+	def __init__(self,
+		decoder: TransformerDecoder,
+		encoder: nn.Module,
+		encoder_token: int,
```

Comment:
why singular?

Reply:
I believe this is a special token like "<img>" or something similar

Reply:
This is the token that represents where encoded embeddings go. I had thought to make all of the FusionModels take in multiple encoders by default but thought it might be better to handle that when we need to support it and update it then.